core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2MetadataAsyncClient.java

```diff
@@ -120,28 +120,33 @@ private CompletableFuture<MetadataResponse> get(String path, RetryPolicyContext
                                   .retriesAttempted(newAttempt)
                                   .exception(SdkClientException.create(error.getMessage(), error))
                                   .build();
-            return scheduledRetryAttempt(() -> get(path, newContext), newContext);
-        }).thenCompose(Function.identity()); // only java 12 has .exceptionallyCompose()
+            scheduledRetryAttempt(() -> get(path, newContext, returnFuture), newContext);
+        });
     }
 
     private CompletableFuture<String> sendAsyncRequest(SdkHttpFullRequest baseRequest) {
-        StringAsyncResponseHandler responseHandler = new StringAsyncResponseHandler();
         SdkHttpContentPublisher requestContentPublisher = new SimpleHttpContentPublisher(baseRequest);
+        StringResponseHandler stringResponseHandler = new StringResponseHandler();
+        TransformingAsyncResponseHandler<String> responseHandler = new AsyncResponseHandler<>(stringResponseHandler,
+                                                                                              Function.identity(),
+                                                                                              new ExecutionAttributes());
+        CompletableFuture<String> future = responseHandler.prepare();
```

Comment:
Nit: There's a lot of futures to handle; can this future have a more descriptive name?